src/WebJobs.Script.Grpc/Server/ExtensionsCompositeEndpointDataSource.cs

```diff
@@ -0,0 +1,300 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Azure.WebJobs.Host.Rpc;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Azure.WebJobs.Script.Grpc
+{
+    /// <summary>
+    /// Endpoint data source which composes all the WebJobs extension data sources together.
+    /// </summary>
+    /// <remarks>
+    /// Implementation is adapted from <see cref="CompositeEndpointDataSource"/>.
+    /// https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/CompositeEndpointDataSource.cs.
+    /// </remarks>
+    internal sealed class ExtensionsCompositeEndpointDataSource : EndpointDataSource, IDisposable
+    {
+        private readonly object _lock = new();
+        private readonly List<EndpointDataSource> _dataSources = new();
+        private readonly IScriptHostManager _scriptHostManager;
+
+        private IServiceProvider _extensionServices;
+        private List<Endpoint> _endpoints;
+        private IChangeToken _consumerChangeToken;
+        private CancellationTokenSource _cts;
+        private List<IDisposable> _changeTokenRegistrations;
+        private bool _disposed;
+
+        public ExtensionsCompositeEndpointDataSource(IScriptHostManager scriptHostManager)
+        {
+            _scriptHostManager = scriptHostManager;
+            _scriptHostManager.ActiveHostChanged += OnHostChanged;
+        }
+
+        /// <inheritdoc />
+        public override IReadOnlyList<Endpoint> Endpoints
+        {
+            get
+            {
+                ThrowIfDisposed();
+                EnsureEndpointsInitialized();
+                return _endpoints;
+            }
+        }
+
+        /// <inheritdoc />
+        public override IChangeToken GetChangeToken()
+        {
+            ThrowIfDisposed();
+            EnsureChangeTokenInitialized();
+            return _consumerChangeToken;
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
```

Comment:
Should this follow the dispose implementation pattern https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/dispose-pattern#basic-dispose-pattern
?

Reply:
The purpose of that pattern is for inheritance. This class is `sealed`, so there are no potential inheritors, and that pattern is not needed. See [CA1063](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1063), which directly mentions using `sealed` as a fix for this.